akka24-conductr-client-lib/src/main/scala/com/typesafe/conductr/clientlib/akka/ControlClient.scala

```diff
@@ -68,71 +71,87 @@ class ControlClient(handler: ConnectionHandler, conductrAddress: URL, apiVersion
    * @return The result as a Future[BundleGetResult]. BundleGetResult is a sealed trait and can be either:
    *         - BundleGetSuccess if the get bundle request has been succeeded. This object contains the bundle id, bundle file, and optionally the config file.
    *         - BundleGetFailure if the get bundle request has been failed. This object contains the HTTP status code and error message.
    */
-  override def getBundle(bundleId: BundleId)(implicit cc: ConnectionContext): Future[BundleGetResult] =
+  override def getBundle(bundleId: BundleId, bundleData: Subscriber[Array[Byte]], configData: Subscriber[Array[Byte]])(implicit cc: ConnectionContext): Future[BundleGetResult] =
     handler.withConnectedRequest(Payload.getBundle(bundleId)) { (responseCode, responseHeader, responseEntity) =>
       import cc.actorMaterializer
       import cc.actorMaterializer.executionContext
 
       def bundleGet: Future[BundleGetResult] = {
 
-        def bundleFileFromResponse(bundleParts: Seq[(String, String, Multipart.FormData.BodyPart)]): BundleFile =
+        def bundleFileFromResponse(bundleParts: Seq[(String, Multipart.FormData.BodyPart)]): String =
           bundleParts match {
-            case Seq(entry) if entry._1 == "bundle" =>
-              val (_, fileName, bodyPart) = entry
-              BundleFile(fileName, toByteArrayPublisher(bodyPart.entity.dataBytes))
+            case Seq((fileName, bodyPart)) if bodyPart.name == "bundle" =>
+              bodyPart.entity.dataBytes
+                .map(_.toArray)
+                .runWith(Sink.fromSubscriber(bundleData))
+
+              fileName
 
             case _ =>
-              throw new InvalidBundleGetResponseBody("Unable to find bundle file in the response body")
+              val error = InvalidBundleGetResponseBody("Unable to find bundle file in the response body")
+
+              Source.failed[Array[Byte]](error)
+                .alsoTo(Sink.fromSubscriber(configData))
+                .runWith(Sink.fromSubscriber(bundleData))
+
+              throw error
           }
 
-        def configFileFromResponse(configParts: Seq[(String, String, Multipart.FormData.BodyPart)]): Option[BundleConfigurationFile] =
-          configParts.find(_._1 == "configuration") match {
-            case Some((_, fileName, bodyPart)) =>
-              Some(BundleConfigurationFile(fileName, toByteArrayPublisher(bodyPart.entity.dataBytes)))
+        def configFileFromResponse(configParts: Seq[(String, Multipart.FormData.BodyPart)]): Option[String] =
+          configParts.find(_._2.name == "configuration") match {
+            case Some((fileName, configPart)) =>
+              configPart.entity.dataBytes
+                .map(_.toArray)
+                .runWith(Sink.fromSubscriber(configData))
+
+              Some(fileName)
 
             case _ =>
+              Source.empty[Array[Byte]]
+                .runWith(Sink.fromSubscriber(configData))
               None
           }
 
         for {
```

Comment:
We could also retrieve the bundle file and bundle configuration in parallel by creating the futures outside of the for-comprehension.